sdk/iot/hub/src/az_iot_hub_client.c

```diff
@@ -50,39 +50,38 @@ AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
   const az_span* const module_id = &(client->_internal.options.module_id);
   const az_span* const user_agent = &(client->_internal.options.user_agent);
 
-  int32_t required_length = az_span_length(client->_internal.iot_hub_hostname)
-      + az_span_length(client->_internal.device_id) + az_span_length(hub_service_api_version) + 1;
-  if (az_span_length(*module_id) > 0)
+  int32_t required_length = az_span_size(client->_internal.iot_hub_hostname)
+      + az_span_size(client->_internal.device_id) + az_span_size(hub_service_api_version) + 1;
+  if (az_span_size(*module_id) > 0)
   {
-    required_length += az_span_length(*module_id) + 1;
+    required_length += az_span_size(*module_id) + 1;
   }
-  if (az_span_length(*user_agent) > 0)
+  if (az_span_size(*user_agent) > 0)
   {
-    required_length += az_span_length(*user_agent) + 1;
+    required_length += az_span_size(*user_agent) + 1;
   }
 
-  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(mqtt_user_name, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_user_name, required_length);
 
-  mqtt_user_name = az_span_copy(mqtt_user_name, client->_internal.iot_hub_hostname);
-  mqtt_user_name = az_span_append_uint8(mqtt_user_name, hub_client_forward_slash);
-  mqtt_user_name = az_span_append(mqtt_user_name, client->_internal.device_id);
+  az_span remainder = az_span_copy(mqtt_user_name, client->_internal.iot_hub_hostname);
+  remainder = az_span_copy_uint8(remainder, hub_client_forward_slash);
+  remainder = az_span_copy(remainder, client->_internal.device_id);
 
-  if (az_span_length(*module_id) > 0)
+  if (az_span_size(*module_id) > 0)
   {
-    mqtt_user_name = az_span_append_uint8(mqtt_user_name, hub_client_forward_slash);
-    mqtt_user_name = az_span_append(mqtt_user_name, *module_id);
+    remainder = az_span_copy_uint8(remainder, hub_client_forward_slash);
+    remainder = az_span_copy(remainder, *module_id);
   }
 
-  mqtt_user_name = az_span_append(mqtt_user_name, hub_service_api_version);
+  remainder = az_span_copy(remainder, hub_service_api_version);
 
-  if (az_span_length(*user_agent) > 0)
+  if (az_span_size(*user_agent) > 0)
   {
-    mqtt_user_name
-        = az_span_append_uint8(mqtt_user_name, *az_span_ptr(hub_client_param_separator_span));
-    mqtt_user_name = az_span_append(mqtt_user_name, *user_agent);
+    remainder = az_span_copy_uint8(remainder, *az_span_ptr(hub_client_param_separator_span));
+    az_span_copy(remainder, *user_agent);
   }
 
-  *out_mqtt_user_name = mqtt_user_name;
+  *out_mqtt_user_name = az_span_slice(mqtt_user_name, 0, required_length);
```

Comment:
Nit: should this be `az_span_init`? They would both do the same thing just curious which we should consistently use.

Reply:
> which we should consistently use.

We should try and avoid using `az_span_init` outside of `az_span.c` and opt-in to use the `az_span_slice` everywhere when it comes to creating a span from another. We should be taking advantage of the precondition validation to catch potential coding bugs.

Similarly, limit uses of `az_span_ptr` where possible.

`az_span_init` should only be used for the first time span creation for certain cases where you happen to have a ptr/length already. Otherwise, use the macros.

Reply:
I wouldn't worry about this at all - all of the public APIs like this one will change to char*, int, int* instead of out_span.